Code refactoring around JobLauncher (see details below)

liyinan926 opened the pull request:
1. Refactored `JobLauncher` to extend `Closeable` and code around its usage to properly close it like other `Closeable`s.
2. Made clear that `JobLauncher` should not be reused and refactored code to somehow reinforce it.
3. Refactored implementation classes of `JobLauncher` to follow the above changes.
4. In `MRJobLauncher`, the `close()` method will now kill the Hadoop job if it has not completed.
5. Other code refactoring.

sahilTakiar commented on gobblin-runtime/src/main/java/gobblin/runtime/JobLauncherFactory.java:
```diff
@@ -37,18 +38,23 @@
+   * @return newly created {@link JobLauncher}
    */
-  public static JobLauncher newJobLauncher(Properties properties)
+  public static @Nonnull JobLauncher newJobLauncher(Properties properties, Properties jobProps)
```
My understanding is that `properties` refers to system properties and `jobProps` is job level properties? Can we rename `properties` to `systemProps`. This way it is more clear to a reader the difference between the two. I think this may require some refactoring in other classes too.

liyinan926 replied:
Fixed.

sahilTakiar commented on gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanJobLauncher.java:
```diff
@@ -40,50 +42,59 @@
 
     // Create a JobLauncher instance depending on the configuration
-    this.jobLauncher = JobLauncherFactory.newJobLauncher(this.properties);
+    this.jobLauncher = this.closer.register(JobLauncherFactory.newJobLauncher(properties, properties));
```
Can you add a comment as to why `JobLauncherFactory.newJobLauncher(Properties, Properties)` has to take in the same `Properties` object for each parameter. I'm guessing it's because Azkaban doesn't make any distinction between properties in `.job` files vs properties in `.properties` files.

liyinan926 replied:
Yes. Comment added.

sahilTakiar commented on gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanJobLauncher.java:
```diff
@@ -40,50 +42,59 @@
 
-    this.properties = new Properties();
-    this.properties.putAll(props);
+    Properties properties = new Properties();
```
Just some food for thought, but should we continue to rely on the Java `Properties` class. It has some weird behavior. Check out the Java Docs: http://docs.oracle.com/javase/7/docs/api/java/util/Properties.html

Specifically the part about the `put` and `putAll` method. The methods directly modify the `HashTable` that backs the `Properties` object, and the methods allow a user to add non-`String` value types.

I've also never been a fan of the `new Properties(Properties defaults)` constructor. It's confusing because it looks like a copy constructor, but in reality it sets property "defaults", which don't get returned by functions such as `keySet()`.

FelixGV replied:
Azkaban and Voldemort both have their own Props class with strongly-typed
getter/setters and other sensical commodities.

Although copy-pasting is evil, it may be a lesser evil than using regular
Java Properties ;)

We should really just stick these things in a lib and re-use them wherever
needed...

-F
On Tue, May 5, 2015 at 18:12 sahilTakiar <notifications@github.com> wrote:

> In gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanJobLauncher.java
> <https://github.com/linkedin/gobblin/pull/120#discussion_r29728730>:
>
> >
> >    public AzkabanJobLauncher(String jobId, Properties props)
> >        throws Exception {
> >      super(jobId, LOG);
> >
> > -    this.properties = new Properties();
> > -    this.properties.putAll(props);
> > +    Properties properties = new Properties();
>
> Just some food for thought, but should we continue to rely on the Java
> Properties class. It has some weird behavior. Check out the Java Docs:
> http://docs.oracle.com/javase/7/docs/api/java/util/Properties.html
>
> Specifically the part about the put and putAll method. The methods
> directly modify the HashTable that backs the Properties object, and the
> methods allow a user to add non-String value types.
>
> I've also never been a fan of the new Properties(Properties defaults)
> constructor. It's confusing because it looks like a copy constructor, but
> in reality it sets property "defaults", which don't get returned by
> functions such as keySet().
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/linkedin/gobblin/pull/120/files#r29728730>.
>

liyinan926 replied:
That's the reason why I am making a copy here.

sahilTakiar commented on gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java:
```diff
@@ -60,36 +58,94 @@
       this.jobHistoryStore = Optional.absent();
     }
+
+    this.jobName = jobProps.getProperty(ConfigurationKeys.JOB_NAME_KEY);
```
Missing `this.`

liyinan926 replied:
Fixed.

sahilTakiar commented on gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java:
```diff
@@ -105,60 +105,125 @@
       LOG.error("Failed to kill the Hadoop MR job " + this.job.getJobID(), ioe);
       throw new JobException("Failed to kill the Hadoop MR job " + this.job.getJobID(), ioe);
+    } finally {
+      cleanUpWorkingDirectory();
```
It may be better to do this in `GobblinOutputCommitter`. When canceling a job via Azkaban, the process only has 5 ms to run the `cancel()` method, before Azkaban does a hard kill on the Java process (e.g. kill -9 [pid]). So it's rarely the case where Gobblin gets to do much (if any) cleanup. Calling `this.job.killJob()` make take a while, and `cleanUpWorkingDirectory` may take a while also.

Also, the `GobblinOutputCommitter` actually depends on the presence of the `mrJobDir`. It actually deserializes the `WorkUnit`s from the directory, and tries to cleanup the writer staging directories for each `WorkUnit`. This work used to be done in Gobblin, but was moved to the output committer due to the 5 ms timeout behavior. Since the output committer logic runs in the AM it has more time to do additional cleanup.

liyinan926 replied:
Yes, we should also do it in `GobblinOutputCommitter`. Also I think it does not hurt in this case to do it in both places to be more defensive.

sahilTakiar replied:
I think that the problem is that `GobblinOutputCommitter` relies on the data in `mrJobDir`. It uses the files in the folder to create `WorkUnit`, and then cleans up the staging data for each `WorkUnit`. Since the OutputCommitter runs in the Application Master, and Gobblin runs on the Azkaban machine, there could be some race condition which causes the OutputCommitter to fail.

sahilTakiar commented on gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java:
```diff
@@ -458,7 +486,7 @@ protected void setup(Context context) {
       } catch (IOException ioe) {
         LOG.error("Failed to setup the mapper task", ioe);
-        return;
+        throw new RuntimeException(ioe);
```
Why not replace above two lines with:

`throw new RuntimException("Failed to setup the mapper task", ioe);`

liyinan926 replied:
Fixed.

sahilTakiar commented on gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java:
```diff
@@ -105,60 +105,125 @@
+      // Wait for the job to complete
+      this.job.waitForCompletion(true);
+
+      if (this.isCancelled) {
```
Is this the safest approach? If there are two threads, where Thread 1 is executing `runWorkUnits()` and Thread 2 is executing `cancelJob()`, then the following scenario may be possible.

Thread 1 -> `this.job.waitForCompletion(true)`
Thread 2 -> `cancelJob()`
Thread 2 -> `cancelJob() -> this.job.killJob()` succeeds
Thread 1 -> Exits from `this.job.waitForCompletion(true)`
Thread 1 -> `if (this.isCancelled)` evaluates to false, and the `jobState` doesn't get set properly
Thread 2 -> `this.isCancelled = true`

liyinan926 replied:
Yes, true. Fixed.

sahilTakiar commented on gobblin-runtime/src/main/java/gobblin/runtime/local/LocalJobLauncher.java:
```diff
@@ -109,35 +122,31 @@ protected void runJob(String jobName, Properties jobProps, JobState jobState, Li
-      jobState.setState(JobState.RunningState.CANCELLED);
-    } else if (jobState.getState() == JobState.RunningState.RUNNING) {
-      jobState.setState(JobState.RunningState.SUCCESSFUL);
+    if (this.isCancelled) {
```
Same race condition as `MRJobLauncher`, except with the `CountDownLatch`

liyinan926 replied:
Fixed.

sahilTakiar commented on gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java:
```diff
@@ -60,36 +58,94 @@
+  protected final JobState jobState;
+  private final JobCommitPolicy jobCommitPolicy;
+  private final boolean jobLockEnabled;
+  private final Optional<JobMetrics> jobMetricsOptional;
```
It may not be necessary, but should we consider having a separate `Job` class that contains `jobProps`, `jobName`, `jobState`, `jobCommitPolicy`, etc. There seems to be a lot of configurations related to a `Job`, so a separate `Job` class may be a better abstraction. Then `AbstractJobLauncher` has a `Job` instance variable, which makes the relationship between a `JobLauncher` and a `Job` clear. A `JobLauncher` **has a** `Job`.

liyinan926 replied:
That's a good point.

liyinan926 replied:
I think the name `JobContext` may be even better.

chavdar commented on gobblin-runtime/src/main/java/gobblin/runtime/JobListener.java:
```diff
@@ -19,7 +19,14 @@
+   * @param jobState a {@link JobState} object
    */
-  public void jobCompleted(JobState jobState);
+  public void onJobCompletion(JobState jobState);
```
JobListenter is a public API. We have to get into the habit of not changing public APIs in a backwards-incompatible manner unless absolutely necessary.

liyinan926 replied:
Yeah, agreed.

chavdar replied:
So, can we leave the old name in this case?

liyinan926 replied:
I would rather keep the new name since currently this interface and all its implementation classes are only used internally, so it's probably the best time to correct the name (`onJobCompletion` sounds more appropriate for the name of a callback method) before it's used publicly.

liyinan926 commented:
Updated 7fd552d6ac5ce5e4f8f4b059946cff198a5c872b.

liyinan926 commented:
Please ignore the last update. I am working on a better way of handling cancellation.

liyinan926 commented:
OK, updated 71b5c7d1d51be8a970223da3461c5233ea346f1e.

liyinan926 commented:
Updated. Major changes are:
1. The way cancellation is handled. AbstractJobLauncher now has a default implementation of cancelJob that simply sets the flag cancellationRequested (which is volatile boolean), which indicates that cancellation is requested. Then the method will wait (with timeout) on a object named cancellationLock on which notify will be called once the cancellation is successfully executed. No explicit synchronization is needed for cancellationRequested since it is volatile boolean. The responsibility of executing the cancellation is done by a single-thread scheduled thread pool executor named `cancellationExecutor`, which periodically checks if the flag `cancellationRequested` is set and calls a method `executeCancellation` to actually execute the cancellation if the flag is set. If the cancellation is successfully executed, another flag `cancellationExecuted` is set so the `cancellationExecutor` will do a noop if the cancellation has already been executed. The execution of cancellation and setting of the `cancellationExecuted` flag, as well as the check of the flag `cancellationExecuted` in the 'JobLauncher' implementations, are synchronized on `cancellationExecutor` (used as a lock).
2. Added a new class JobContext that contains all context information of a job. This simplifies AbstractJobLauncher.
3. Added a new method `shutdownExecutorService` in `ExecutorUtils` and used it in the code base.

sahilTakiar commented:
LGTM. Although I'm not entirely sure on the design of the `cancel()` method. Perhaps @chavdar can take a look?

chavdar commented on gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java:
```diff
@@ -116,97 +132,69 @@ public AbstractJobLauncher(Properties properties)
       LOG.warn("No work units have been created for job " + jobId);
-      source.shutdown(jobState);
-      unlockJob(jobName, jobLockOptional);
+      unlockJob();
```
unlockJob() should be in a finally section. Having to call unlockJob in every exit branch is very error-prone.

chavdar commented on gobblin-utility/src/main/java/gobblin/util/ExecutorsUtils.java:
```diff
@@ -65,4 +67,29 @@ public static ThreadFactory newThreadFactory(Optional<Logger> logger, Optional<S
+      }
+    } catch (InterruptedException ie) {
+      executorService.shutdownNow();
+      Thread.currentThread().interrupt();
```
i think it is better to rethrow ie, to preserve the original stack trace.

liyinan926 commented:
@chavdar, updated to address your comments.

chavdar commented on gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java:
```diff
@@ -116,139 +132,103 @@ public AbstractJobLauncher(Properties properties)
   @Override
-  @SuppressWarnings("unchecked")
-  public void launchJob(Properties jobProps, JobListener jobListener)
+  public void cancelJob(JobListener jobListener)
```
* What is the current contract of the cancelJob() API? Is it synchronous or asynchronous? Are we guaranteeing to always call the listener? 

Currently, it seems like sometimes the call is executed synchronously and sometimes, it is not. The listener is called back only if the call is executed synchronously but I think that kind of defeats the purpose of having a listener (aka callback). If the call is synchronous, I can execute whatever I want after the cancelJob() call.
I think the current form of the API suggests asynchronous cancellation. This means that the implementation should always return immediately and always call the callback regardless if there is cancellation in progress or not. We can of course have different contract but it should be described very carefully in the javadocs.

* How does cancelJob() relate to the state property in the job state? My intuition is that cancelJob() should set the state to CANCELLED at the end of the run but that does not seem to be the case currently. The state property is managed independently which I think is error prone. I may be missing some interaction, in which case it may be useful to document it properly. 

* Overall, I think we need a global lock on state transitions, not only cancellationLock and we need to manage those more carefully. Currently, close() seems to be closing the cancellation executor regardless of whether a cancellation is in progress. Similarly, I think it is also possible to cancel a job after it has already finished (successfully or not). I find this counter-intuitive. I think having a state machine diagram with the transition will be helpful.

* I think the current implementation with synchronization on cancellationExecutor and cancellationLock is a bit complicated. I think having a single state transition lock (see above) and appropriate condition variables might be cleaner.

liyinan926 replied:
I agree with the above points and what you are concerned about here, and it sounds like we need a thorough design on job state/lifecycle management before moving forward with any further changes around job cancellation handling. This is certainly beyond the scope of this pull request, which aims at providing a better default implementation around job cancellation than the previous one. Are you OK with the current default implementation?

liyinan926 commented:
Updated a0add8c.

chavdar commented:
I understand that streamlining the state transitions will require some work so I agree we should do that later but I think we should decide on the cancelJob() contract and clean up the implementation. I think this is fairly self-contained work.

liyinan926 commented:
@sahilTakiar updated c6cf621. Can you review?
